### PR TITLE
Simplify creation of NodeEntry instances and use emplace_back

### PR DESCRIPTION
**Pull request by larroy**
## Description ##

Reduce copying of shared_ptr. In a microbenchmark is 10% faster. But the main motivation is simplifying creation of nnvm graph on operators.

    Optimize move semantics of NodeEntry
    https://github.com/dmlc/tvm/pull/2576
    Making copies of shared_ptr is more expensive than moving.
    This PR reduces lock contention by using move semantics in NNVM nodes
    making also more convenient to construct NodeEntry classes in the code
    due to the added ctors.
    
    Update NDarray with NodeEntry constructors and refine initializer lists.
    
    Sync gradient.cc with tvm

imperative.cc will be addressed in a different refactor.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by vandanavk** [comment]
@mxnet-label-bot add [pr-work-in-progress]

**Comment by ptrendx** [comment]
Does this change improve the time of scheduling launches to the engine?

**Comment by larroy** [comment]
@ptrendx thanks for your comment, would still need to measure

**Comment by abhinavs95** [comment]
@larroy Could you please fix the CI failures?

**Comment by piyushghai** [comment]
@larroy Is this PR still WIP ?

**Comment by larroy** [comment]
@ptrendx suggestions on measurements?

File: .gitmodules [hunk 1]
```diff
@@ -22,7 +22,7 @@
 	branch = master
 [submodule "3rdparty/tvm"]
 	path = 3rdparty/tvm
-	url = https://github.com/dmlc/tvm
+	url = https://github.com/larroy/tvm
```
Comment (apeforest) [inline .gitmodules]: please remember to update this when dmlc PR is merged.

File: dev_menu.py [hunk 2]
```diff
@@ -99,21 +99,21 @@ def create_virtualenv(venv_exe, pyexe, venv) -> None:
         logging.warn("Skipping creation of virtualenv")
         return
     check_call([venv_exe, '-p', pyexe, venv])
-    activate_this_py = os.path.join(venv, 'bin', 'activate_this.py')
-    # Activate virtualenv in this interpreter
-    exec(open(activate_this_py).read(), dict(__file__=activate_this_py))
-    check_call(['pip', 'install', '--upgrade','--force-reinstall', '-e', 'python'])
-    check_call(['pip', 'install', '-r', 'tests/requirements.txt'])
+    # TODO: Activate virtualenv in this interpreter, this is not working as it is.
```
Comment (apeforest) [inline dev_menu.py]: This change seems to be unrelated to this PR. Can we make the PR focused and maybe create a separte PR for this change?

**Comment by gigasquid** [comment]
@larroy I was trying to build the branch to look at the Clojure stuff - but I ran into the same problem as CI 

```
fatal: reference is not a tree: e031b39c5e1d4468b7db96f1db39dbb97e4beef8

Unable to checkout 'e031b39c5e1d4468b7db96f1db39dbb97e4beef8' in submodule path '3rdparty/tvm'

script returned exit code 1
```
ping me when this is resolved and I'll repro / look at the Clojure error

File: .gitmodules (hunk 1 above)
Comment (larroy) [reply .gitmodules]: ok

File: dev_menu.py (hunk 2 above)
Comment (larroy) [reply dev_menu.py]: Yes, is still marked as [Don't merge], it's WIP.

**Comment by larroy** [comment]
@gigasquid now the clojure issue is isolated. Is the only test failing. Would appreciate your help.

**Comment by larroy** [comment]
@mxnet-label-bot add [Backend]

**Comment by larroy** [comment]
@gigasquid can you please help with Clojure?

**Comment by gigasquid** [comment]
@larroy I have a PR to fix the Clojure failures https://github.com/larroy/mxnet/pull/2

**Comment by larroy** [comment]
@mxnet-label-bot update [Backend,pr-awaiting-review]

File: src/executor/infer_graph_attr_pass.cc [hunk 3]
```diff
@@ -628,7 +628,8 @@ nnvm::Graph InferShapeAttr(nnvm::Graph &&ret,
     }
     if (dispatch_mode_name) {
       for (size_t i = node_start; i < node_end; i++) {
-        if (dispatch_modes[i] == DispatchMode::kUndefined) ++num_unknown;
+        if (dispatch_modes[i] == DispatchMode::kUndefined)
```
Comment (apeforest) [inline src/executor/infer_graph_attr_pass.cc]: nit: add braces around if block.

File: include/mxnet/ndarray.h [hunk 4]
```diff
@@ -82,7 +82,8 @@ class MKLDNNMemory;
 class NDArray {
  public:
   /*! \brief default constructor */
-  NDArray() {
+  NDArray()
+    : entry_(nullptr) {
```
Comment (apeforest) [inline include/mxnet/ndarray.h]: doesn't the NodeEntry constructor require 3 arguments?

File: src/operator/custom/custom.cc [hunk 5]
```diff
@@ -238,14 +238,14 @@ std::vector<nnvm::NodeEntry> Gradient(
 
   std::vector<nnvm::NodeEntry> ret;
   for (size_t i = 0; i < params.num_args; ++i) {
-    ret.emplace_back(nnvm::NodeEntry{g, static_cast<uint32_t>(i), 0});
+    ret.emplace_back(g, static_cast<uint32_t>(i), 0);
```
Comment (apeforest) [inline src/operator/custom/custom.cc]: do you still need to cast to uint32_t?

File: src/nnvm/legacy_op_util.cc [hunk 6]
```diff
@@ -340,17 +341,15 @@ inline std::vector<NodeEntry> OpPropGradient(
...
   }
   // attach no gradient node to forbid gradient on aux_state
   if (prop.aux_states.size() != 0) {
-    NodePtr ng = Node::Create();
-    ng->attrs.op = Op::Get("_NoGradient");
-    ng->attrs.name = "NoGradient";
-    for (uint32_t i = 0; i < prop.aux_states.size(); ++i) {
-      in_grad.emplace_back(NodeEntry{ng, 0, 0});
+    for (size_t i = 0; i < prop.aux_states.size(); ++i) {
+      in_grad.emplace_back(Node::Create(Op::Get("_NoGradient"), "NoGradient"), 0, 0);
```
Comment (apeforest) [inline src/nnvm/legacy_op_util.cc]: Node::Create() returns a NodePtr, but here it requires a NodeEntry. Is it correct here?

File: include/mxnet/ndarray.h (hunk 4 above)
Comment (larroy) [reply include/mxnet/ndarray.h]: I refactored it.

File: src/operator/custom/custom.cc (hunk 5 above)
Comment (larroy) [reply src/operator/custom/custom.cc]: Yes otherwise there's a warning.

**Comment by larroy** [comment]
@szha what would you suggest?

**Comment by larroy** [comment]
@szha the initial motivation is to lean out node creation and for correctness since there are difficult to catch bugs depending on Node initialization.

**Comment by larroy** [comment]
@szha  Performance is the same, so motivation is correctness and ease of graph building:


Tested in a tight loop:

```
TEST(NodeTestX, NodeTest) {
    using namespace nnvm;
    using namespace std;
    using namespace std::chrono;
    vector<nnvm::NodeEntry> v;
    nnvm::NodePtr ng = nnvm::Node::Create();
    ng->attrs.op = Op::Get("_zeros_without_dtype");
    ng->attrs.name = "zeros_without_dtype";
#if 0
    high_resolution_clock::time_point t1 = high_resolution_clock::now();
    for(size_t i = 0; i < 10000000; ++i) {
        v.push_back(NodeEntry{ng, 0, 0});
    }
    high_resolution_clock::time_point t2 = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>( t2 - t1 ).count();
    cout << duration;
#endif
#if 1
    auto t1 = high_resolution_clock::now();
    for(size_t i = 0; i < 10000000; ++i) {
        v.emplace_back(ng, 0, 0);
    }
    auto t2 = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>( t2 - t1 ).count();
    cout << duration;
#endif
}
```


```
piotr@ec2 cpu:0: ~/mxnet [master]> build/tests/mxnet_unit_tests --gtest_filter="NodeTest*"

Note: Google Test filter = NodeTest*
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from NodeTestX
[ RUN      ] NodeTestX.NodeTest
1659863[       OK ] NodeTestX.NodeTest (1918 ms)
[----------] 1 test from NodeTestX (1918 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (1919 ms total)
[  PASSED  ] 1 test.
```

```

Note: Google Test filter = NodeTest*
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from NodeTestX
[ RUN      ] NodeTestX.NodeTest
1717020[       OK ] NodeTestX.NodeTest (1985 ms)
[----------] 1 test from NodeTestX (1985 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (1985 ms total)
[  PASSED  ] 1 test.
```

**Comment by larroy** [comment]
Moving a shared pointer is faster than copying. When using move in the microbenchmark is 10% faster.  Overall is not much and in the big picture won't make a difference, but these small inefficiencies compound together and make the code in general slower in a way that you can't profile as it's scatered all over.

What changes exactly are you suggesting? to the title of the PR? is not clear to me, please help clarify.

It can also cause contention between threads due to the atomic lock.
 
```
Note: Google Test filter = NodeTest*
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from NodeTestX
[ RUN      ] NodeTestX.NodeTest
1655650[       OK ] NodeTestX.NodeTest (1804 ms)
[----------] 1 test from NodeTestX (1804 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (1804 ms total)
[  PASSED  ] 1 test.
```

**Comment by szha** [comment]
The information in the PR suggests this patch is premature optimization as it's not making any difference in speed. If any of the symptoms can be captured in a test, please do so. Otherwise, if it's a usability improvement, then documenting the suggested use would be helpful.

**Comment by larroy** [comment]
I just provided a microbenchmark where you can see it's 10% faster. Usually in C++ we try to avoid doing more work than necessary such as copying a shared_ptr when is not necessary. Do you disagree with that? It creates inefficiencies compounded that don't show in a profiler.   It also makes it easier and less verbose to add to an nnvm graph.

**Comment by larroy** [comment]
I don't think is premature optimization, is generally agreed that is better to emplace_back rather than push_back and create additional copies, it also makes less verbose to add nodes. Why you call it a premature optimization? What exact changes you are suggesting?

**Comment by szha** [comment]
I'm worried that it would be premature because of the lack of the measurable improvements in the problem space what mxnet concerns.

**Comment by larroy** [comment]
Would you be willing to merge since it simplifies creating graphs on operators if you don't find the performance improvement justified?

https://github.com/apache/incubator-mxnet/pull/14992
https://github.com/apache/incubator-mxnet/pull/14613

**Comment by larroy** [comment]
@szha I changed the title according to your request, are we good to merge?

**Comment by szha** [comment]
Thanks for documenting the change. This PR needs one more rebase.

**Comment by larroy** [comment]
@szha Any more comments, or are we ready to go?
